modules/apps/frontend-js/frontend-js-importmaps-extender/src/main/java/com/liferay/frontend/js/importmaps/extender/internal/servlet/taglib/JSImportMapsCache.java

```diff
@@ -138,53 +188,132 @@ private Map<String, String> _getScopedImportMapsValues(Long companyId) {
 		return _scopedImportMapsValuesMap.get(companyId);
 	}
 
-	private void _writeImports(
-			Map<Long, String> globalImportMapsValues, Writer writer)
+	private void _writeDynamicImports(
+			Map<Long, DynamicJSImportMapsContributor>
+				dynamicJSImportMapsContributors,
+			HttpServletRequest httpServletRequest, StringBuilder sb)
 		throws IOException {
 
 		boolean first = true;
 
-		for (String value : globalImportMapsValues.values()) {
-			if (!first) {
-				writer.write(StringPool.COMMA);
+		for (DynamicJSImportMapsContributor dynamicJSImportMapsContributor :
+				dynamicJSImportMapsContributors.values()) {
+
+			CharArrayWriter charArrayWriter = new CharArrayWriter();
+
+			dynamicJSImportMapsContributor.writeGlobalImports(
+				httpServletRequest, charArrayWriter);
+
+			if (charArrayWriter.size() == 0) {
+				continue;
 			}
-			else {
+
+			if (first) {
 				first = false;
+
+				if (!sb.isEmpty()) {
+					sb.append(StringPool.COMMA);
+				}
+			}
+			else {
+				sb.append(StringPool.COMMA);
 			}
 
-			writer.write(value);
+			sb.append(charArrayWriter);
 		}
 	}
 
-	private void _writeScopes(
-			Map<String, String> scopedImportMapsValues, Writer writer)
+	private void _writeDynamicScopes(
+			Map<Long, DynamicJSImportMapsContributor>
+				dynamicJSImportMapsContributors,
+			HttpServletRequest httpServletRequest, StringBuilder sb)
 		throws IOException {
 
 		boolean first = true;
 
+		for (DynamicJSImportMapsContributor dynamicJSImportMapsContributor :
+				dynamicJSImportMapsContributors.values()) {
+
+			CharArrayWriter charArrayWriter = new CharArrayWriter();
+
+			dynamicJSImportMapsContributor.writeScopedImports(
+				httpServletRequest, charArrayWriter);
+
+			if (charArrayWriter.size() == 0) {
+				continue;
+			}
+
+			if (first) {
+				first = false;
+
+				if (!sb.isEmpty()) {
+					sb.append(StringPool.COMMA);
+				}
+			}
+			else {
+				sb.append(StringPool.COMMA);
+			}
```

Comment:
I think this is for dangling commas in a json(I might be wrong), but I don't think browsers care about that any more, do they?

Reply:
I don't know for sure, but extra commas are not technically valid JSON so better be safe.